Userland/Libraries/LibAudio/FlacWriter.cpp

```diff
@@ -0,0 +1,613 @@
+{
+}
+
+FlacWriter::~FlacWriter()
+{
+    if (m_state != WriteState::FullyFinalized)
+        MUST(finalize());
+}
+
+ErrorOr<void> FlacWriter::finalize()
+{
+    if (m_state == WriteState::FullyFinalized)
+        return Error::from_string_view("File is already finalized"sv);
+
+    // TODO: Write missing sample data instead of discarding it.
+
+    if (m_state == WriteState::HeaderUnwritten)
+        TRY(finalize_header_format());
+    {
+        // 1 byte metadata block header + 3 bytes size + 2*2 bytes min/max block size
+        TRY(m_stream->seek(m_streaminfo_start_index + 8, AK::SeekMode::SetPosition));
+        BigEndianOutputBitStream bit_stream { MaybeOwned<Stream> { *m_stream } };
+        TRY(bit_stream.write_bits(m_min_frame_size, 24));
+        TRY(bit_stream.write_bits(m_max_frame_size, 24));
+        TRY(bit_stream.write_bits(m_sample_rate, 20));
+        TRY(bit_stream.write_bits(m_num_channels - 1u, 3));
+        TRY(bit_stream.write_bits(m_bits_per_sample - 1u, 5));
+        TRY(bit_stream.write_bits(m_sample_count, 36));
+        TRY(bit_stream.align_to_byte_boundary());
+    }
+
+    // TODO: Write the audio data MD5 to the header.
+
+    m_stream->close();
+
+    m_state = WriteState::FullyFinalized;
+    return {};
+}
+
+ErrorOr<void> FlacWriter::finalize_header_format()
+{
+    if (m_state != WriteState::HeaderUnwritten)
+        return Error::from_string_view("Header format is already finalized"sv);
+    TRY(write_header());
+    m_state = WriteState::FormatFinalized;
+    return {};
+}
+
+ErrorOr<void> FlacWriter::set_num_channels(u8 num_channels)
+{
+    if (m_state != WriteState::HeaderUnwritten)
+        return Error::from_string_view("Header format is already finalized"sv);
+    if (num_channels > 8)
+        return Error::from_string_view("FLAC doesn't support more than 8 channels"sv);
+
+    m_num_channels = num_channels;
+    return {};
+}
+
+ErrorOr<void> FlacWriter::set_sample_rate(u32 sample_rate)
+{
+    if (m_state != WriteState::HeaderUnwritten)
+        return Error::from_string_view("Header format is already finalized"sv);
+
+    m_sample_rate = sample_rate;
+    return {};
+}
+
+ErrorOr<void> FlacWriter::set_bits_per_sample(u16 bits_per_sample)
+{
+    if (m_state != WriteState::HeaderUnwritten)
+        return Error::from_string_view("Header format is already finalized"sv);
+    if (bits_per_sample < 8 || bits_per_sample > 32)
+        return Error::from_string_view("FLAC only supports bits per sample between 8 and 32"sv);
+
+    m_bits_per_sample = bits_per_sample;
+    return {};
+}
+
+ErrorOr<void> FlacWriter::write_header()
+{
+    TRY(m_stream->write_until_depleted(flac_magic.bytes()));
+    m_streaminfo_start_index = TRY(m_stream->tell());
+
+    ByteBuffer data;
+    // STREAMINFO is always exactly 34 bytes long.
+    TRY(data.try_resize(34));
+    BigEndianOutputBitStream header_stream { TRY(try_make<FixedMemoryStream>(data.bytes())) };
+
+    // Duplication on purpose; min/max block size.
```

Comment:
Maybe instead comment "min block size" on the line encoding the min size, and "max block size" on the line encoding the max size?